src/aks-preview/azext_aks_preview/custom.py

```diff
@@ -2687,104 +2727,219 @@ def _ensure_container_insights_for_monitoring(cmd, addon):
-        "contentVersion": "1.0.0.0",
-        "parameters": {
-            "workspaceResourceId": {
-                "type": "string",
-                "metadata": {
-                    "description": "Azure Monitor Log Analytics Resource ID"
-                }
-            },
-            "workspaceRegion": {
-                "type": "string",
-                "metadata": {
-                    "description": "Azure Monitor Log Analytics workspace region"
+    if not remove_monitoring:
+        resources = cf_resources(cmd.cli_ctx, subscription_id)
+        from azure.core.exceptions import HttpResponseError
+        try:
+            resource = resources.get_by_id(
+                workspace_resource_id, '2015-11-01-preview')
+            location = resource.location
+        except HttpResponseError as ex:
+            raise ex
+
+    if aad_route:
+        cluster_resource_id = f"/subscriptions/{cluster_subscription}/resourceGroups/{cluster_resource_group_name}/providers/Microsoft.ContainerService/managedClusters/{cluster_name}"
+        dataCollectionRuleName = f"DCR-{workspace_name}"
+        dcr_resource_id = f"/subscriptions/{subscription_id}/resourceGroups/{resource_group}/providers/Microsoft.Insights/dataCollectionRules/{dataCollectionRuleName}"
+        from azure.cli.core.util import send_raw_request
+        from azure.cli.core.profiles import ResourceType
+
+        if create_dcr:
+            # first get the association between region display names and region IDs (because for some reason
+            # the "which RPs are available in which regions" check returns region display names)
+            region_names_to_id = {}
+            # retry the request up to two times
+            for _ in range(3):
+                try:
+                    location_list_url = f"https://management.azure.com/subscriptions/{subscription_id}/locations?api-version=2019-11-01"
+                    r = send_raw_request(cmd.cli_ctx, "GET", location_list_url)
+
+                    # this is required to fool the static analyzer. The else statement will only run if an exception
+                    # is thrown, but flake8 will complain that e is undefined if we don't also define it here.
+                    e = None
+                    break
+                except CLIError as e:
+                    pass
+            else:
+                # This will run if the above for loop was not broken out of. This means all three requests failed
+                raise e
+            json_response = json.loads(r.text)
+            for region_data in json_response["value"]:
+                region_names_to_id[region_data["displayName"]] = region_data["name"]
+
+            # check if region supports DCRs and DCR-A
+            for _ in range(3):
+                try:
+                    feature_check_url = f"https://management.azure.com/subscriptions/{subscription_id}/providers/Microsoft.Insights?api-version=2020-10-01"
```

Comment:
At present, are there corresponding methods in SDK that can call those REST API? If so, why not consider using the Python SDK methods instead of directly requesting REST?

Reply:
I'm not sure if there are python SDK methods to check for feature availability. I directly used rest because I couldn't figure out how to call "https://management.azure.com/subscriptions/{subscription_id}/providers/Microsoft.Insights?api-version=2020-10-01" and "https://management.azure.com/subscriptions/{subscription_id}/locations?api-version=2019-11-01" with the Python SKD.

Reply:
At present, there is no document for Python SDK. It depends on whether there is a swagger defined for those REST
It seems that there is no support from SDK methods, so this PR can be merge first